groth16/src/ext_wit.rs

```diff
@@ -27,77 +26,125 @@ pub async fn h<
-    // King receives shares of p, q, w
-    let h_share = if let (Some(p_shares), Some(q_shares), Some(w_shares)) =
-        (received_p_shares, received_q_shares, received_w_shares)
-    {
-        let unpack_shares = |v| {
-            let mut s1 = cfg_into_iter!(transpose(v))
-                .flat_map(|x| pp.unpack(x))
-                .collect::<Vec<_>>();
-            // swap each ith element with l*i+t element
-            // TODO: Guru should help explaining this
-            for i in 0..m {
-                s1.swap(i, i * pp.l + pp.t);
-            }
-            s1
-        };
-        let mut p_eval = unpack_shares(p_shares);
-        let mut q_eval = unpack_shares(q_shares);
-        let mut w_eval = unpack_shares(w_shares);
-
-        p_eval.truncate(m);
-        q_eval.truncate(m);
-        w_eval.truncate(m);
-
-        // King do the final step of multiplication.
-        let h = cfg_into_iter!(p_eval)
-            .zip(q_eval)
-            .zip(w_eval)
-            .map(|((p, q), w)| p.mul(q).sub(w))
-            .collect::<Vec<_>>();
-        // pack and send to parties
-        let h_shares = transpose(pack_vec(&h, pp));
-        net.recv_from_king(Some(h_shares), CHANNEL0).await?
-    } else {
-        net.recv_from_king(None, CHANNEL0).await?
+    let coset_dom = domain.get_coset(F::GENERATOR).unwrap();
+
+    let a_coeff_fut = d_ifft(
+        qap_share.a,
+        true,
+        &domain,
+        coset_dom.coset_offset(),
+        pp,
+        net,
+        CHANNEL0,
+    );
+    let b_coeff_fut = d_ifft(
+        qap_share.b,
+        true,
+        &domain,
+        coset_dom.coset_offset(),
+        pp,
+        net,
+        CHANNEL1,
+    );
+    let c_coeff_fut = d_ifft(
+        qap_share.c,
+        true,
+        &domain,
+        coset_dom.coset_offset(),
+        pp,
+        net,
+        CHANNEL2,
+    );
+
+    let (a_coeff, b_coeff, c_coeff) =
+        tokio::try_join!(a_coeff_fut, b_coeff_fut, c_coeff_fut)?;
+
+    let a_eval_fut = d_fft(a_coeff, true, &domain, pp, net, CHANNEL0);
+    let b_eval_fut = d_fft(b_coeff, true, &domain, pp, net, CHANNEL1);
+    let c_eval_fut = d_fft(c_coeff, true, &domain, pp, net, CHANNEL2);
+
+    // evaluations of a, b, c over the coset
+    let (a_eval, b_eval, c_eval) =
+        tokio::try_join!(a_eval_fut, b_eval_fut, c_eval_fut)?;
+
+    // compute (ab-c)/z
+    let vanishing_polynomial_over_coset = domain
+        .evaluate_vanishing_polynomial(F::GENERATOR)
+        .inverse()
+        .unwrap();
+
+    let h_eval = cfg_into_iter!(a_eval)
+        .zip(b_eval)
+        .zip(c_eval)
+        .map(|((a, b), c)| (a * b - c) * vanishing_polynomial_over_coset)
+        .collect::<Vec<_>>();
+
+    // run coset_ifft to get back coefficients of h
+    let h_coeff_fut = d_ifft(
+        h_eval,
+        false,
+        &domain,
+        coset_dom.coset_offset_inv(),
+        pp,
+        net,
+        CHANNEL0,
+    );
+
+    let h_coeff = tokio::try_join!(h_coeff_fut)?.0;
```

Comment:
If this only one future, you can just do a `.await?` directly, there is no joining needed here.
```suggestion
    // run coset_ifft to get back coefficients of h
    let h_coeff = d_ifft(
        h_eval,
        false,
        &domain,
        coset_dom.coset_offset_inv(),
        pp,
        net,
        CHANNEL0,
    ).await?;
```